### PR TITLE
Dev 1419

### PR DESCRIPTION
**Pull request by AntonZelenin**

### CODE REVIEW

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 1]
```diff
@@ -39,10 +39,12 @@
 import javax.ws.rs.core.MultivaluedMap;
 import javax.ws.rs.core.StreamingOutput;
 import javax.ws.rs.core.Response;
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
+import java.io.*;
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java]: Please, remove start import "*"

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 2]
```diff
@@ -103,6 +105,12 @@ public void destroy() {
     public void write(Batch batch) throws StageException {
         try {
             Iterator<Record> records = batch.getRecords();
+            Record lastRecord = getLastRecord(batch);
+            if (lastRecord != null && !conf.agentOffsetUrl.equals("")) {
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java:109]: `!conf.agentOffsetUrl.equals("")` should be replaced with `!"".equals(conf.agentOffsetUrl)`
Not sure if `conf.agentOffsetUrl` can throw NullPointerException, so better to be safe

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 1 above)
Comment (kiamran) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: http://www.javadude.com/posts/20040522-import-on-demand-is-evil/

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 3]
```diff
@@ -139,6 +147,40 @@ public void write(Batch batch) throws StageException {
...
+        Iterator<Record> iter = batch.getRecords();
+        Record record = null;
+        while (iter.hasNext()) {
+            record = iter.next();
+        }
+        return record;
+    }
+
+    private void sendOffsetToAgent(String offset, String pipelineId) throws Exception {
+        URLConnection con = new URL(conf.agentOffsetUrl + pipelineId).openConnection();
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java]: I there any simpler method to make post request using `streamsets-datacollector-httpcommonlib` ?

Comment (kiamran) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: Also, i think we need to set some kind of timeout in case agent is not responsible

Comment (kiamran) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: do we also need to close `HttpURLConnection`?

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 4]
```diff
@@ -139,6 +147,40 @@ public void write(Batch batch) throws StageException {
...
+        }
+        if (http.getResponseCode() != 200) {
+            try(BufferedReader br = new BufferedReader(
+                    new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8))) {
+                StringBuilder response = new StringBuilder();
+                String responseLine;
+                while ((responseLine = br.readLine()) != null) {
+                    response.append(responseLine.trim());
+                }
+                throw new Exception(response.toString());
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java]: maybe should be replaced with 

` new OnRecordErrorException()` or something not generic like `Exception`

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 5]
```diff
@@ -103,6 +105,12 @@ public void destroy() {
     public void write(Batch batch) throws StageException {
         try {
             Iterator<Record> records = batch.getRecords();
+            Record lastRecord = getLastRecord(batch);
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java:110]: `records` could be passed instead of `batch `

Comment (AntonZelenin) [reply src/main/java/anodot/stage/destination/AnodotTarget.java:108]: write method is a part of the interface and it accepts batch

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 4 above)
Comment (AntonZelenin) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: well, yeah, it just wasn't meant to be handled so I put Exception here

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 5 above)
Comment (AntonZelenin) [reply src/main/java/anodot/stage/destination/AnodotTarget.java:110]: yes, but as for me it doesn't make much difference

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 4 above)
Comment (kiamran) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: maybe add more context than?
Something like `failed to saved agent timestamp. HTTP error ${response_code}: ${response_body}`

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 6]
```diff
@@ -103,6 +110,12 @@ public void destroy() {
...
+            if (lastRecord != null && !conf.agentOffsetUrl.equals("")) {
+                String pipelineId = lastRecord.get().getValueAsMap().get("tags").getValueAsMap().get("pipeline_id").getValueAsList().get(0).getValueAsString();
```
Comment (dariakharlan) [inline src/main/java/anodot/stage/destination/AnodotTarget.java:112]: Maybe there is a better way to get pipeline id? Even if we have to set additional parameters for that

**Review by dariakharlan — CHANGES_REQUESTED** [review]

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 6 above)
Comment (AntonZelenin) [reply src/main/java/anodot/stage/destination/AnodotTarget.java:112]: will be left as it is for now, as discussed

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 7]
```diff
@@ -36,13 +38,13 @@
...
+import javax.ws.rs.core.MultivaluedHashMap;
 import javax.ws.rs.core.MultivaluedMap;
 import javax.ws.rs.core.StreamingOutput;
 import javax.ws.rs.core.Response;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java]: please remove star import

File: src/main/java/anodot/stage/destination/AnodotTarget.java [hunk 8]
```diff
@@ -139,6 +147,33 @@ public void write(Batch batch) throws StageException {
...
+        MultivaluedMap<String, Object> requestHeaders = new MultivaluedHashMap<>();
+        String contentType = HttpStageUtil.getContentType(requestHeaders, DataFormat.JSON);
+        Response response = builder.method(String.valueOf(HttpMethod.POST), Entity.entity(String.format("{\"offset\": \"%s\"}", offset).getBytes(StandardCharsets.UTF_8), contentType));
+
+        if (response.getStatus() < 200 || response.getStatus() >= 300) {
+            String responseEntity = response.readEntity(String.class);
+            response.close();
+            throw new Exception("Failed to save agent offset, response: " + responseEntity);
+        }
+        response.close();
```
Comment (kiamran) [inline src/main/java/anodot/stage/destination/AnodotTarget.java]: will not be executed when the exception will be thrown before.
You can use try-finally

**Review by kiamran — CHANGES_REQUESTED** [review]

File: src/main/java/anodot/stage/destination/AnodotTarget.java (hunk 7 above)
Comment (AntonZelenin) [reply src/main/java/anodot/stage/destination/AnodotTarget.java]: damn, where did it come from?? stupid auto imports

**Review by kiamran — APPROVED** [review]
Please, do not forget to use squash-and-merge  :)

**Review by dariakharlan — APPROVED** [review]

**Pull request merged**